diff --git a/events.go b/events.go
index 87ad7d9..9bc5ce5 100644
--- a/events.go
+++ b/events.go
@@ -260,6 +260,8 @@ type MessageDeleteBulk struct {
 	Messages  []string `json:"ids"`
 	ChannelID string   `json:"channel_id"`
 	GuildID   string   `json:"guild_id"`
+
+	BeforeDelete map[string]*Message `json:"-"`
 }
 
 // WebhooksUpdate is the data for a WebhooksUpdate event
diff --git a/state.go b/state.go
index 2eeabd8..2987c84 100644
--- a/state.go
+++ b/state.go
@@ -931,7 +931,15 @@ func (s *State) OnInterface(se *Session, i interface{}) (err error) {
 		}
 	case *MessageDeleteBulk:
 		if s.MaxMessageCount != 0 {
+			t.BeforeDelete = make(map[string]*Message)
 			for _, mID := range t.Messages {
+				// shadowing of err avoided here since this isn't an error that needs to be handled
+				old, ex := s.Message(t.ChannelID, mID)
+				if ex == nil {
+					oldCopy := *old
+					t.BeforeDelete[mID] = &oldCopy
+				}
+
 				s.messageRemoveByID(t.ChannelID, mID)
 			}
 		}
